### ISSUE TITLE
yarn install --production does not install dependencies that are transitive dependencies of a dependency and a dev dependency

### ISSUE THREAD

**Issue by dreid** [yarnpkg/yarn#2073]
**What is the current behavior?**

When a package is depended on by both a dependency and a devDependency, the devDependency visibility takes precedence when installing into production, this results in shared dependencies not being installed in production.

**If the current behavior is a bug, please provide the steps to reproduce.**
https://github.com/dreid/yarn-dev-dep-failure

`yarn install --production` does not install function-bind

`npm install --production` does install function-bind

**What is the expected behavior?**

Production builds should install dependencies that are visible from production dependencies.

**Please mention your node.js, yarn and operating system version.**

```
❯ yarn --version
0.17.9
❯ node -v
v7.1.0
❯ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.12.1
BuildVersion:	16B2555
```

**Comment by bestander** [comment]
Making it higher priority, thanks

**Comment by itamararjuan** [comment]
@bestander are you fixing this yourself ?
was thinking of helping solve this as it's totally preventing me from moving some of my services to yarn.

**Comment by bestander** [comment]
@itamararjuan, I would totally appreciate if you fix this.
It may take some time for me to get to this one.
Please don't forget to include a test

**Comment by itamararjuan** [comment]
@bestander on it!

**Comment by kittens** [comment]
Looks like this was fixed with #2116, it required significant refactor to how we do package visibility. Sorry @itamararjuan, didn't realise that the issue I was fixing was so pathalogical and extended to this case too.

**Comment by itamararjuan** [comment]
hey sebastien
no worries thanks for doing so much for everybody. this is an awesome
project. if help is needed dont hesitate to ping and I'll do my best to
help with other issues :)
